qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py

```diff
@@ -0,0 +1,128 @@
+        self.set_function_string('blah')
+        return self.wait_for_modal()
+
+    def set_function_string_linear(self):
+        self.set_function_string('name=LinearBackground')
+        return self.wait_for_true(lambda: self.widget.sizeOfFunctionsGroup() == 3)
+
+    def set_function_string(self, text):
+        box = self.get_active_modal_widget()
+        box.setTextValue(text)
+        QMetaObject.invokeMethod(box, 'accept', Qt.QueuedConnection)
+
+    def test_find_peaks_no_workspace(self):
+        yield self.start_setup_menu()
+        m = self.get_menu('menu_Setup')
+        self.test.assertTrue(m.isVisible())
+        yield self.start_find_peaks()
+        box = self.get_active_modal_widget()
+        self.test.assertTrue(isinstance(box, QMessageBox))
+        self.test.assertEqual(box.text(), 'Workspace name is not set')
+        box.close()
+
+    def test_load_from_string_blah(self):
+        yield self.start_setup_menu()
+        yield self.start_manage_setup()
+        yield self.start_load_from_string()
+        yield self.set_function_string_blah()
+        box = self.get_active_modal_widget()
+        self.test.assertEqual(box.text(),
+                              "Unexpected exception caught:\n\nError in input string to FunctionFactory\nblah")
+        box.close()
+
+    def test_load_from_string_lb(self):
+        yield self.start_setup_menu()
+        yield self.start_manage_setup()
+        yield self.start_load_from_string()
+        yield self.set_function_string_linear()
+        a = self.widget.getFittingFunction()
+        self.test.assertEqual(a, 'name=LinearBackground,A0=0,A1=0')
+        self.test.assertEqual(self.widget.sizeOfFunctionsGroup(), 3)
+
+    def test_copy_to_clipboard(self):
+        self.widget.loadFunction('name=LinearBackground,A0=0,A1=0')
+        yield self.start_setup_menu()
+        yield self.start_manage_setup()
+        QApplication.clipboard().clear()
+        self.trigger_action('action_CopyToClipboard')
+        yield self.wait_for_true(lambda: QApplication.clipboard().text() != '')
+        self.test.assertEqual(QApplication.clipboard().text(), 'name=LinearBackground,A0=0,A1=0')
+
+    def test_clear_model(self):
+        self.widget.loadFunction('name=LinearBackground,A0=0,A1=0')
+        self.test.assertEqual(self.widget.sizeOfFunctionsGroup(), 3)
+        yield self.start_setup_menu()
+        yield self.start_manage_setup()
+        self.trigger_action('action_ClearModel')
+        yield self.wait_for_true(lambda: self.widget.sizeOfFunctionsGroup() == 2)
+        self.test.assertEqual(self.widget.sizeOfFunctionsGroup(), 2)
+
+
+@unittest.skipIf(skip_on_ubuntu(), "Popups don't show on ubuntu with python 2. Unskip when switched to xvfb")
+class TestModalTester(unittest.TestCase):
+
+    def tearDown(self):
+        import time
+        time.sleep(0.2)
+
+    def test_find_peaks_no_workspace(self):
+        TestFitPropertyBrowser(self).run('test_find_peaks_no_workspace', pause=0.)
```

Comment:
Something worrying I noticed is that this will always 'pass' even if the test it runs fails. Small example:

```python
class TestSmallExample(GuiTestBase):
    def __init__(self, test):
        super(TestSmallExample, self).__init__(GenericDialog)
        self.test = test

    def test_fail(self):
        self.test.fail("Forced fail")


class TestSmallExampleTester(unittest.TestCase):
    def test_fail_does_not_fail(self):
        TestSmallExample(self).run('test_fail', pause=0.)
```

Gives me output
```
1871: Running tests...
1871: ----------------------------------------------------------------------
1871: Traceback (most recent call last):
1871:   File "C:\Users\qbr77747\dev\m\source\qt\python\mantidqt\utils\qt\test\run_test_app.py", line 158, in run_script
1871:     return fun(widget)
1871:   File "C:\Users\qbr77747\dev\m\source\qt\python\mantidqt\utils\qt\test\test_window.py", line 20, in __call__
1871:     return getattr(self, self.call_method)()
1871:   File "C:/Users/qbr77747/dev/m/source/qt/python/mantidqt/widgets/test/test_algorithmdialog.py", line 84, in test_fail
1871:     self.test.fail("Forced fail")
1871:   File "C:\Users\qbr77747\dev\m\source\external\src\ThirdParty\lib\python2.7\lib\unittest\case.py", line 410, in fail
1871:     raise self.failureException(msg)
1871: AssertionError: Forced fail
1871: .
1871: ----------------------------------------------------------------------
1871: Ran 1 test in 0.063s
1871:
1871: OK
1871:
1871: Generating XML reports...
2/2 Test #1871: mantidqt_qt4.test_algorithmdialog.test_algorithmdialog ...   Passed    7.12 sec

The following tests passed:
        mantidqt_qt5.test_algorithmdialog.test_algorithmdialog
        mantidqt_qt4.test_algorithmdialog.test_algorithmdialog

100% tests passed, 0 tests failed out of 2
```

Any idea how to address this? I tried getting the return value from `TestFitPropertyBrowser(self).run`, which in turn runs `app._exec()` down the call chain, but it seems to always be 0 regardless of whether the internal test passes or fails.

Reply:
Yes. The error gets caught by `run_script` and the traceback is printed out.

Reply:
I'll fix it.